tests/helpers/test_config_validation.py

```diff
@@ -481,26 +481,303 @@ def test_datetime():
     schema('2016-11-23T18:59:08')
 
 
-def test_deprecated(caplog):
-    """Test deprecation log."""
-    schema = vol.Schema({
+@pytest.fixture
+def schema():
+    """Create a schema used for testing deprecation."""
+    return vol.Schema({
         'venus': cv.boolean,
-        'mars': cv.boolean
+        'mars': cv.boolean,
+        'jupiter': cv.boolean
     })
+
+
+def test_deprecated_with_no_optionals(caplog, schema):
+    """
+    Test deprecation behaves correctly when optional params are None.
+
+    Expected behavior:
+        - Outputs the appropriate deprecation warning if key is detected
+        - Processes schema without changing any values
+        - No warning or difference in output if key is not provided
+    """
     deprecated_schema = vol.All(
         cv.deprecated('mars'),
         schema
     )
 
-    deprecated_schema({'venus': True})
-    # pylint: disable=len-as-condition
+    test_data = {'mars': True}
+    output = deprecated_schema(test_data.copy())
+    assert len(caplog.records) == 1
+    assert caplog.records[0].name == __name__
+    assert ("The 'mars' option (with value 'True') is deprecated, "
+            "please remove it from your configuration.") in caplog.text
+    assert test_data == output
+
+    caplog.clear()
+    assert len(caplog.records) == 0
+
+    test_data = {'venus': True}
+    output = deprecated_schema(test_data.copy())
+    assert len(caplog.records) == 0
+    assert test_data == output
+
+
+def test_deprecated_with_replacement_key(caplog, schema):
+    """
+    Test deprecation behaves correctly when only a replacement key is provided.
+
+    Expected behavior:
+        - Outputs the appropriate deprecation warning if key is detected
+        - Processes schema moving the value from key to replacement_key
+        - Processes schema changing nothing if only replacement_key provided
+        - No warning if only replacement_key provided
+        - No warning or difference in output if neither key nor
+            replacement_key are provided
+    """
+    deprecated_schema = vol.All(
+        cv.deprecated('mars', replacement_key='jupiter'),
+        schema
+    )
+
+    test_data = {'mars': True}
+    output = deprecated_schema(test_data.copy())
+    assert len(caplog.records) == 1
+    assert ("The 'mars' option (with value 'True') is deprecated, "
+            "please replace it with 'jupiter'.") in caplog.text
+    assert {'jupiter': True} == output
+
+    caplog.clear()
+    assert len(caplog.records) == 0
+
+    test_data = {'jupiter': True}
+    output = deprecated_schema(test_data.copy())
+    assert len(caplog.records) == 0
+    assert test_data == output
+
+    test_data = {'venus': True}
+    output = deprecated_schema(test_data.copy())
+    assert len(caplog.records) == 0
+    assert test_data == output
+
+
+def test_deprecated_with_invalidation_version(caplog, schema):
```

Comment:
We should patch the home assistant version in this test. Otherwise we'll get another result in the future when we pass the version used in the test.

Reply:
Agreed, good catch.